mautic/mautic
# Mcrypt to OpenSSL

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | None
| Deprecations? | Mcrypt

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
- Mcrypt usage from EncryptionHelper moved to new class McryptCipher
- New class OpenSSLCipher created
- EncryptionHelper accepts N classes implementing ISymmetricCipher (order determine priority)
- EncryptionHelper::encrypt Encrypted using by top supported class
- EncryptionHelper::decrypt Try supported classes by priority until one succeed
- EncryptionHelper::decrypt now has 2nd optional parameter $mainDecryptOnly (default = false) which force method to return false without trying other class if main supported class decryption fails
- EncryptionHelper covered by UnitTests
- AbstractIntegration::decryptApiKeys now has 2nd optional parameter $mainDecryptOnly (default = false) used in call of EncryptionHelper.:decrypt
- AbstractIntegration::decryptApiKeys now return empty array in case of false decryption of any key.
- AbstractIntegration::getDecryptedApiKeys first try decrypt keys by main decrypt only.
** In case of success everything is OK.
** In case of empty array return, keys are decrypted with all possible methods, re-encrypted using AbstractIntegration::encryptAndSetApiKeys, Entity is persisted and flushed.

[//]: # ( As applicable: )

#### Steps to test this PR:
Test without OpenSSL (if app will use deprecated Mcrypt in case OpenSSL is not supported):
1. Checkout to master.mcryptToOpenSSL
2. Try to change secret token at any social plugin settings that supports it
3. Everything should work

Test with OpenSSL but without provided access to /dev/random or /dev/urandom by open_basedir
- Test by previous scenario (app should use deprecated Mcrypt again)

Test with OpenSSL enabled and with access to /dev/random or /dev/urandom
1. Checkout to master
2. Backup any api_keys attribute with secret token from MySQL database plugin setting
3. Checkout to master.mcryptToOpenSSL (this branch)
4. Do anything in Mautic app
5. Check same api_keys again in MySQL database - it should be different than previous backup
6. Plugin authorization should be still valid and working

#### List deprecations along with the new alternative:
1. Mcrypt library deprecated - OpenSSL is alternative

#### List backwards compatibility breaks:
None
